theano/tensor/nnet/opt.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import print_function
+
+import theano
+from theano import tensor
+
+from theano.gof import local_optimizer
+from theano.tensor.nnet import (
+    CorrMM, CorrMM_gradInputs, CorrMM_gradWeights)
+from theano.tensor.nnet.abstract_conv2d import (BaseAbstractConv2d, AbstractConv2d,
+                                                AbstractConv2d_gradWeights,
+                                                AbstractConv2d_gradInputs)
+from theano.tensor.opt import register_specialize_device
+from theano.tensor import TensorType
+
+
+### Corrmm opt
+@local_optimizer([AbstractConv2d])
+def local_abstractconv_gemm(node):
+    if not isinstance(node.op, AbstractConv2d):
+        return None
+    img, kern = node.inputs
+    if (not isinstance(img.type, TensorType) or
+        not isinstance(kern.type, TensorType)):
+        return None
+
+    # need to flip the kernel if necessary
+    if node.op.filter_flip:
+        kern = kern[:, :, ::-1, ::-1]
+    # By default use CorrMM
+    rval = CorrMM(border_mode=node.op.border_mode,
+                  subsample=node.op.subsample)(img, kern)
+
+    return [rval]
+
+@local_optimizer([AbstractConv2d_gradWeights])
+def local_abstractconv_gradweight_gemm(node):
+    if not isinstance(node.op, AbstractConv2d_gradWeights):
+        return None
+    img, topgrad, shape = node.inputs
+    if (not isinstance(img.type, TensorType) or \
+        not isinstance(topgrad.type, TensorType)):
+        return None
+
+    rval = CorrMM_gradWeights(border_mode=node.op.border_mode,
+                              subsample=node.op.subsample)(img, topgrad, shape)
+    if node.op.filter_flip:
+        rval = rval[:, :, ::-1, ::-1]
+    rval = tensor.patternbroadcast(rval, node.outputs[0].broadcastable)
+    return [rval]
+
+@local_optimizer([AbstractConv2d_gradInputs])
+def local_abstractconv_gradinputs_gemm(node):
+    if not isinstance(node.op, AbstractConv2d_gradInputs):
+        return None
+    kern, topgrad, shape = node.inputs
+    if (not isinstance(kern.type, TensorType) or \
+        not isinstance(topgrad.type, TensorType)):
+        return None
+
+    if node.op.filter_flip:
+        kern = kern[:, :, ::-1, ::-1]
+
+    rval =  CorrMM_gradInputs(border_mode=node.op.border_mode,
+                              subsample=node.op.subsample)(kern, topgrad,
+                                                           shape)
+    return [rval]
+
+# Register CPU convolution implementation
+# They are tried in a specific order so we can control
+# which ones take precedence over others.
+abstractconv_groupopt = theano.gof.optdb.LocalGroupDB()
+abstractconv_groupopt.__name__ = "abstractconv_opts"
+#TODO all below
+register_specialize_device(abstractconv_groupopt, 'gpu', 'fast_compile')
+
+# The GEMM-based convolution comes last to catch all remaining cases.
+# It can be disabled by excluding 'conv_gemm'.
+conv_groupopt.register('local_abstractconv_gemm', local_abstractconv_gemm, 30,
```

Comment:
The groupopt here should have the same name that you created above. I have no real preference.